### DISCUSSION TITLE
Can we make the backend like NodeJs to use Authelia APIs to login on behalf of user and forward the cookies back to client ?

### DISCUSSION THREAD

**Discussion by vm-pried** [authelia/authelia#7197] [Q&A] [answered]
Hi,

We are looking for using our Login WebPage , obtain the credentials and call Authelia API to login 

As part of this response, can we take the session cookies etc given from Authelia and pass it back to the client along with the Token.

Is this possible ? If so any pointers on which APIs would we need to call and the order for the firstfactor auth

Any help would be greatly appreciated.

**Comment by glime-ai** [comment]
Please note this response was created using generative AI leveraging previously available information from the repository. This information may be helpful however it is our strong recommendation that you carefully consider the information before blindly applying it.

---

Hello,

To achieve the integration you're describing with Authelia, where you handle the login on your webpage and then interact with Authelia's API to authenticate and pass session cookies and tokens back to the client, you will need to follow a specific sequence of API calls. Here’s a general outline of the steps involved:

1. **Collect Credentials**: On your custom login page, collect the user's credentials.

2. **Call Authelia's First Factor API**: Use the credentials collected to make a POST request to Authelia’s `/api/firstfactor` endpoint. This will authenticate the user's first factor (usually a password).

   Example POST request:
   ```bash
   curl -X POST https://authelia.example.com/api/firstfactor -d '{"username":"user", "password":"password"}'
   ```

3. **Handle the Response**: If the authentication is successful, Authelia will return a session cookie in the response headers. You need to capture this cookie.

4. **Pass Session Cookie to Client**: Forward the session cookie to the client's browser. This cookie is essential for maintaining the authenticated session with Authelia.

5. **Use Token for Further Actions**: If you are using tokens (like JWTs from OpenID Connect), ensure that after the first factor authentication, you proceed with the necessary token retrieval steps (this might involve additional endpoints depending on your setup with Authelia).

It’s important to ensure that all communications are done over HTTPS to secure the credentials and session information being transmitted.

For detailed API documentation and more specific guidance, you might need to refer to Authelia's official documentation or source code, as the exact endpoints and parameters might depend on the version of Authelia you are using and any custom configurations.

If you encounter any specific issues or need further customization, consulting the Authelia community on GitHub or their official support channels might provide more tailored assistance.

This comment was generated by [Glime](https://glimelab.ai/?utm_source=6fsHkcyufjQLaB&utm_medium=github).

**Comment by james-d-elliott** [comment]
You can technically do this by copying the Set-Cookie header as per normal web semantics, this however is not an intended use case so YMMV. You can see all of the API endpoints and what they do by viewing `https://auth.example.com/api` where `https://auth.example.com` is your authelia URL.

I'd recommend using something like https://github.com/auth0/express-openid-connect to perform logins instead.

**Comment by james-d-elliott** [comment, marked as answer]
Here's an Express.js example I drummed up with theoretical values: https://www.authelia.com/integration/openid-connect/expressjs/

**Comment by vm-pried** [comment]
Thanks a lot @james-d-elliott . 

Will try this and get back. Appreciate your quick help on this
